prov/util/src/util_cq.c

```diff
@@ -38,58 +38,67 @@
 
 #define UTIL_DEF_CQ_SIZE (1024)
 
-/* Caller must hold `cq_lock` */
-int ofi_cq_write_overflow(struct util_cq *cq, void *context, uint64_t flags, size_t len,
-			  void *buf, uint64_t data, uint64_t tag, fi_addr_t src)
+
+static void ofi_cq_insert_aux(struct util_cq *cq,
+			      struct util_cq_aux_entry *entry)
 {
-	struct util_cq_oflow_err_entry *entry;
+	entry->cq_slot = ofi_cirque_tail(cq->cirq);
+	entry->cq_slot->flags = UTIL_FLAG_AUX;
+	slist_insert_tail(&entry->list_entry, &cq->aux_queue);
+}
 
-	assert(ofi_cirque_isfull(cq->cirq));
+/* Caller must hold 'cq lock' */
+int ofi_cq_write_overflow(struct util_cq *cq, void *context, uint64_t flags,
+			  size_t len, void *buf, uint64_t data, uint64_t tag,
+			  fi_addr_t src)
+{
+	struct util_cq_aux_entry *entry;
+
+	FI_DBG(cq->domain->prov, FI_LOG_CQ, "writing to CQ overflow list\n");
+	assert(ofi_cirque_freecnt(cq->cirq) == 1);
 
 	if (!(entry = calloc(1, sizeof(*entry))))
 		return -FI_ENOMEM;
 
-	entry->parent_comp = ofi_cirque_tail(cq->cirq);
-	entry->parent_comp->flags |= UTIL_FLAG_OVERFLOW;
-
 	entry->comp.op_context = context;
 	entry->comp.flags = flags;
 	entry->comp.len = len;
 	entry->comp.buf = buf;
 	entry->comp.data = data;
 	entry->comp.tag = tag;
-
+	entry->comp.err = 0;
 	entry->src = src;
-	slist_insert_tail(&entry->list_entry, &cq->oflow_err_list);
 
+	ofi_cq_insert_aux(cq, entry);
 	return 0;
 }
 
-int ofi_cq_write_error(struct util_cq *cq,
-		       const struct fi_cq_err_entry *err_entry)
+/* Caller must hold 'cq lock' */
+int ofi_cq_insert_error(struct util_cq *cq,
+			const struct fi_cq_err_entry *err_entry)
 {
-	struct util_cq_oflow_err_entry *entry;
-	struct fi_cq_tagged_entry *comp;
+	struct util_cq_aux_entry *entry;
 
 	assert(err_entry->err);
-
 	if (!(entry = calloc(1, sizeof(*entry))))
 		return -FI_ENOMEM;
 
 	entry->comp = *err_entry;
-	cq->cq_fastlock_acquire(&cq->cq_lock);
-	slist_insert_tail(&entry->list_entry, &cq->oflow_err_list);
 
-	if (OFI_UNLIKELY(ofi_cirque_isfull(cq->cirq))) {
-		comp = ofi_cirque_tail(cq->cirq);
-		comp->flags |= (UTIL_FLAG_ERROR | UTIL_FLAG_OVERFLOW);
-		entry->parent_comp = ofi_cirque_tail(cq->cirq);
-	} else {
-		comp = ofi_cirque_tail(cq->cirq);
-		comp->flags = UTIL_FLAG_ERROR;
+	ofi_cq_insert_aux(cq, entry);
+	if (ofi_cirque_freecnt(cq->cirq) > 1)
```

Comment:
This doesn't work.  We have to call ofi_cirque_commit, or a reader will never see that the entry is valid.